### ISSUE TITLE
Debug_traceBlockByHash returning "insufficient funds" error on block 19888327

### ISSUE THREAD

**Issue by Enigmatic331** [ethereum/go-ethereum#29800]
#### System information

Geth version: geth.exe version 1.14.3-stable-ab48ba42
CL client & version: beacon-chain version Prysm/v5.0.3/38f208d70dc95b12c08403f5c72009aaa10dfe2f. Built at: 2024-04-04 18:48:31+00:00
OS & Version: Windows

#### Expected behaviour
Debug_traceBlockByHash should return the trace dump for block 19888327

#### Actual behaviour
A tracing failed message was returned instead

```
C:\Users\Administrator>curl http://localhost:8545 -m 300 -X POST --header "Content-type: application/json" --data "{\"jsonrpc\":\"2.0\", \"method\":\"debug_traceBlockByHash\", \"params\":[\"0xcc1d35533ea26392bfa91944b3f6d1f446581b05f5f4b35dea93d558b08d353c\", {\"tracer\":\"callTracer\"}], \"id\":1}"
{"jsonrpc":"2.0","id":1,"error":{"code":-32000,"message":"tracing failed: insufficient funds for gas * price + value: address 0xc79d40C897FbD912a7cD29E318a0D41F579793f6 have 2469004274997060 want 3963014826093176"}}
```

If I independently trace the two "out" transactions on this block for the address [0xc79d40C897FbD912a7cD29E318a0D41F579793f6 ](https://etherscan.io/address/0xc79d40C897FbD912a7cD29E318a0D41F579793f6), it seems to return results correctly.

```
C:\Users\Administrator>curl http://localhost:8545 -m 300 -X POST --header "Content-type: application/json" --data "{\"jsonrpc\":\"2.0\", \"method\":\"debug_traceTransaction\", \"params\":[\"0xefa4345e670ef386c0c1e076625c1a7627df192adf5ed8b7f5a8ec27c14ccdfd\", {\"tracer\":\"callTracer\"}], \"id\":1}"
{"jsonrpc":"2.0","id":1,"result":{"from":"0xc79d40c897fbd912a7cd29e318a0d41f579793f6","gas":"0x927c0","gasUsed":"0x60126","to":"0x5f6ae08b8aeb7078cf2f96afb089d7c9f51da47d","input":"0xba1f77e80000000000000000000000000000000000000000000000000000000000000000","calls":[{... etc
```

```
C:\Users\Administrator>curl http://localhost:8545 -m 300 -X POST --header "Content-type: application/json" --data "{\"jsonrpc\":\"2.0\", \"method\":\"debug_traceTransaction\", \"params\":[\"0xb3552b1f6e4c97349ca45ae34bb831c04c571fd4fa2110481d4aae851e8a0331\", {\"tracer\":\"callTracer\"}], \"id\":1}"
{"jsonrpc":"2.0","id":1,"result":{"from":"0xc79d40c897fbd912a7cd29e318a0d41f579793f6","gas":"0xa410","gasUsed":"0x5208","to":"0x742c5ce550ec2e1a78b338904978ec9004eb065f","input":"0x","value":"0xc8ed706a30328","type":"CALL"}}
```

Seems to be affecting more than 1 of our nodes, potentially other blocks as well as we received some reports about indexed traces showing a transaction failing, however when the tx is traced individually with debug_traceTransaction, no execution reverted seemed to be found.

**Comment by holiman** [comment]
That block contains three transactions from that account

![Screenshot 2024-05-17 at 13-52-33 Address 0xc79d40C897FbD912a7cD29E318a0D41F579793f6 Etherscan](https://github.com/ethereum/go-ethereum/assets/142290/b3a07b3b-3c6c-400b-9638-5340cbfc77a8)

- [`0xb3552b1f6e4c97349ca45ae34bb831c04c571fd4fa2110481d4aae851e8a0331`](https://etherscan.io/tx/0xb3552b1f6e4c97349ca45ae34bb831c04c571fd4fa2110481d4aae851e8a0331)
- [`0xefa4345e670ef386c0c1e076625c1a7627df192adf5ed8b7f5a8ec27c14ccdfd`](https://etherscan.io/tx/0xefa4345e670ef386c0c1e076625c1a7627df192adf5ed8b7f5a8ec27c14ccdfd)
- [`0xd5fd51583eab8fdbb885b017683ee9e5d6b2154103261c49a996cfc542b3dc23`](https://etherscan.io/tx/0xd5fd51583eab8fdbb885b017683ee9e5d6b2154103261c49a996cfc542b3dc23)

**Comment by Enigmatic331** [comment]
The last tx could also run through debug_traceTransaction, though I didn't raise it as it didn't seem to concern 0xc79 as a sender 🙈 - Assuming that the `insufficient funds for gas * price + value: address 0xc79d40C897FbD912a7cD29E318a0D41F579793f6 have 2469004274997060 want 3963014826093176` error targets address 0xc79 sending a transaction to a contract/EOA.

```
C:\Users\Administrator>curl http://localhost:8545 -m 300 -X POST --header "Content-type: application/json" --data "{\"jsonrpc\":\"2.0\", \"method\":\"debug_traceTransaction\", \"params\":[\"0xd5fd51583eab8fdbb885b017683ee9e5d6b2154103261c49a996cfc542b3dc23\", {\"tracer\":\"callTracer\"}], \"id\":1}"
{"jsonrpc":"2.0","id":1,"result":{"from":"0x742c5ce550ec2e1a78b338904978ec9004eb065f","gas":"0xa410","gasUsed":"0x5208","to":"0xc79d40c897fbd912a7cd29e318a0d41f579793f6","input":"0x","value":"0x15bc4cf4757ac0","type":"CALL"}}
```

**Comment by holiman** [comment]
So, looking at the `IN` one state changes: 
![Screenshot 2024-05-17 at 14-04-12 Ethereum Transaction Hash (Txhash) Details Etherscan](https://github.com/ethereum/go-ethereum/assets/142290/54ac3d75-d5a4-44ec-9889-030149065f88)

Before that, the account has nothing. So it's not just a case of "the block-tracing doesn't commit the transaction state changes and executes every trace on the block prestate", because then the error would say `have 0 want...`

**Comment by Enigmatic331** [comment]
Ahhh you're right... Now where did the number 2469004274997060 came from...

**Comment by mtbitcoin** [comment]
Another Block# 19888387 with the similar if you need an extra sample 19888387

**Comment by s1na** [comment]
Are you using the hash scheme or path scheme? If you first synced your node before v1.14.0 and didn't explicitly specify `--state.scheme=path` then it is hash scheme. v1.14.0 flips the default to `path` but only if you do a fresh sync.

**Comment by Enigmatic331** [comment]
We're using hash for our nodes (especially for this one where we are running archive nodes) 🙏

**Comment by quickchase** [comment]
Yeah, archive nodes have to be hash, path isn't supported, but this happens on either, I believe

**Comment by holiman** [comment]
I've been investigating this a bit, haven't found any smoking gun. The `callTracer` is a 'native' tracer, so it doesn't go into the more complicated parallel prestate-execution path, but only the simpler step by step tracing of transactions. 

I have found one thing that looks quite odd, https://github.com/ethereum/go-ethereum/blob/master/eth/tracers/api.go#L985 , where we use a vmconfig with no basefee checking `vm.Config{Tracer: tracer.Hooks, NoBaseFee: true}`, and a mostly empty txcontext with e.g. zero gasprice `vm.TxContext{GasPrice: big.NewInt(0)}`. I don't see how that could be correct.

**Comment by holiman** [comment]
It seems, that earlier, it generated the txContext from the `msg`, but it was changed to the 'empty' form here: 

https://github.com/ethereum/go-ethereum/commit/064f37d6f67a012eea0bf8d410346fb1684004b4#diff-bd288cd51c1874c7e65ce0d35047c0a98366479c8924606c9efc577f6a98f75cL947

@s1na any thoughts why?

**Comment by celochoi** [comment]
There seems to be no problem with `1.13.15`.
Does this issue only occur in 1.14 and above?

**Comment by holiman** [comment]
@s1na the setting of `NoBaseFee: true` will have consequences. Later on, during `vm.NewEVM`: 
```golang
	if config.NoBaseFee {
		if txCtx.GasPrice.BitLen() == 0 {
			blockCtx.BaseFee = new(big.Int)
		}
		if txCtx.BlobFeeCap != nil && txCtx.BlobFeeCap.BitLen() == 0 {
			blockCtx.BlobBaseFee = new(big.Int)
		}
	}
```
That is: we will set the block ctx fields to zero. Thus overwriting the fields that we constructed earlier,  in `eth/tracers/api.go`: 
```
		blockCtx  = core.NewEVMBlockContext(block.Header(), api.chainContext(ctx), nil)
```

This will affect the opcodes `BASEFEE` and `BLOBBASEFEE`. 
If anyone can provide me with the full opcode by opcode output of https://etherscan.io/vmtrace?txhash=0xefa4345e670ef386c0c1e076625c1a7627df192adf5ed8b7f5a8ec27c14ccdfd, that'd be greatly appreciated. I'd like to see what opcodes are executed.

**Comment by pengin7384** [comment]
@holiman 
Did you mean `debug_traceTransaction` response?

Request
```
{
    "method": "debug_traceTransaction",
    "params": [
       "0xefa4345e670ef386c0c1e076625c1a7627df192adf5ed8b7f5a8ec27c14ccdfd"
    ],
    "id": 1,
    "jsonrpc": "2.0"
}
```

Response
[response.json](https://github.com/ethereum/go-ethereum/files/15377485/response.json)

**Comment by quickchase** [comment]
Was the above generated with 1.13 or 1.14? I have 1.13 archive nodes and I have Erigon, Nethermind, and Besu archive nodes, if you want any of those, lemme know 😄

**Comment by holiman** [comment]
Yep, 

It does invoke `BASEFEE`

![Screenshot from 2024-05-20 15-43-17](https://github.com/ethereum/go-ethereum/assets/142290/9040d2e5-fb6b-4184-ad6d-9138c0bb7321)


The value is `0x15818a10a`, and later on does some `LT` checks and jumps depending on it's value

![Screenshot from 2024-05-20 15-44-31](https://github.com/ethereum/go-ethereum/assets/142290/1cb5122d-af3d-4cd5-a8ac-a582e2860ef7)

This is most likely the culprit

**Comment by pengin7384** [comment]
@quickchase It was generated with `geth 1.13.15`.

**Comment by Mazzika1** [comment]
Okay

**Comment by s1na** [comment]
This should have been fixed via #29811. I would appreciate if someone can use latest master and confirm they don't see the issue anymore.

**Comment by s1na** [comment]
Great, closing then. Release coming up next week!
